### PR TITLE
Let vanilla try to handle Container interactions before checking for item handlers

### LINKED ISSUE #1750: Redstone comparators reading inventory changes do not always cause updates to nearby redstone components
**Minecraft Version:** 1.21.3

**NeoForge Version:** 21.3.11-beta

**Logs:** No log needed

**Steps to Reproduce:**
 1. Build this Image
 2. Put 1 item into the chest (e.g. a chest item). Put 5 items into the hopper (e.g. hopper item, stack size 1, hopper item is stackable to 64)
 3. Activate and deactivate the lever

**Description of issue:**
The behavior with neoforge is different from vanilla behavior, breaking certain redstone contraptions. Neoforge redstone comparators seem to only send updates to nearby redstone components when their signal strength actually changes. In vanilla, each inventory update causes the comparator to send updates. Even a player clicking into an empty slot does.

In the setup, the deviation can be observed during step 3:
The piston does not extend or retract when neoforge is installed.
When using vanilla, the piston extends and retracts together with the lever power.

### LINKED ISSUE #1770: Hoppers try to transfer items out twice per tick if the first attempt fails
**Minecraft Version:** 1.21.4

**NeoForge Version:** 21.4.22-beta

**Steps to Reproduce:**
Image

World download:
[HopperTest.zip](https://github.com/user-attachments/files/18135825/HopperTest.zip)
Demo Video:
Cannot post a video here, but demonstration video on the neoforge discord: [1836 ticks (91 seconds) on vanilla vs 1072 ticks (53 seconds) on neoforge](https://discord.com/channels/313125603924639766/852298000042164244/1317329746010312755) The video shows a measurement of the timing. It is not necessary to understand the issue, but may be helpful.

 1. Place a hopper facing into multiple inventory entities (in the world download: 19 chest carts and one chest boat, chest boat sits on a soul sand behind the rail)
 2. Make sure all but one entity inventory is filled entirely (in the world download: chest carts full, chest boat empty)
 3. Put items into the hopper and check how long the transfer takes on average. Compare this to the average time on vanilla. (World download: 86 seconds for 64 items in vanilla, 58 seconds for 64 items on neoforge)

**Description of issue:**
Neoforge's hopper item insertion logic is broken. Neoforge adds a "hook" at the start of the insertion code that duplicates the insertion code in case the insertion attempt fails. For successful attempts, it cancels the execution of the vanilla insertion code. In defail, the hook returns false for unsuccessful insertion attempts and the vanilla attempt is only cancelled when true is returned. (This is implemented differently for the extraction hook.)
Item transfers to inventory entities work by uniformly choosing a random inventory entity that intersects the 1x1x1 volume at the hopper output.
The chance of picking the chest boat in the example is 1/20.
After not choosing the chest boat (19/20 chance), the item transfer attempt fails.
With neoforge a second attempt takes place: Another 1/20 chance to pick the chest boat and successfully transfer.
This significantly speeds up the transfer to the chest boat from ~86 seconds to ~58 seconds for 64 items.

The fix should probably include always cancelling the vanilla transfer code or if possible, avoiding to duplicate the transfer code in the first place.

### PR DESCRIPTION
**Pull request by Technici4n**
Move our `IItemHandler` interaction hooks after vanilla checks for `Container` interactions. This gives vanilla control over interactions with `Container`s. Fixes #1750. Fixes #1770. (Did not test the latter but I assume it gets fixed).
As a side effect, this greatly simplifies our hooks since we don't need to handle vanilla edge cases such as hopper - hopper interactions having special interactions when it comes to cooldowns.

## Breaking change
Modders need to be careful that their `Container`s will be interacted with as `Container`s before an item handler is even attempted to be found. I'd suggest not implementing `Container` directly on block entities to avoid this problem. Nevertheless, this counts as a breaking change.

## Lookup order
From highest priority to lowest priority: `Container block > IItemHandler block > Container entity > IItemHandler entity`.

### CODE REVIEW

**Comment by 2No2Name** [comment]
Confirmed to fix https://github.com/neoforged/NeoForge/issues/1750
Also the measurements for https://github.com/neoforged/NeoForge/issues/1770
(1671 for vanilla, 1201 for Neo w/o fix, 1805 for Neo w/ fix) posted on discord suggest that this PR fixes https://github.com/neoforged/NeoForge/issues/1770 as well.

**Comment by sciwhiz12** [comment]
> Currently, `Container` entities are preferred over `IItemHandler` blocks. What do people think about that?
> 
> Current order: `Container block > Container entity > IItemHandler block > IItemHandler entity`.
> 
> I think that it would be preferable to check in this order: `Container block > IItemHandler block > Container entity > IItemHandler entity`.

How difficult would that be to implement? I can see what you mean by that being preferrable -- ensuring that blocks are always checked before entities in either case -- but I wonder how the impl. look like if you have to interweave the item handler check between container blocks and entities.

**Comment by Technici4n** [comment]
I managed to implement it locally and am planning to update the PR tomorrow.

**Comment by Technici4n** [comment]
PR updated with a more modder-friendly lookup order. :smile: @sciwhiz12

File: src/main/java/net/neoforged/neoforge/items/VanillaInventoryCodeHooks.java [hunk 1]
```diff
@@ -25,15 +24,11 @@
 public class VanillaInventoryCodeHooks {
     /**
      * Copied from TileEntityHopper#captureDroppedItems and added capability support
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/items/VanillaInventoryCodeHooks.java]: Might as well update the referenced method in the comments. Same suggestion for the rest of the comments in this class, insofar as updating references to methods or classes; up to you if you want to pursue rewriting the comments in general.

File: patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch [hunk 2]
```diff
@@ -1,22 +1,88 @@
...
++            return net.neoforged.neoforge.items.VanillaInventoryCodeHooks.extractHook(p_155554_, containerOrHandler.itemHandler());
+         } else {
+             boolean flag = p_155554_.isGridAligned()
+                 && blockstate.isCollisionShapeFullBlock(p_155553_, blockpos)
+@@ -368,6 +_,8 @@
+         return p_155590_.getEntitiesOfClass(ItemEntity.class, aabb, EntitySelector.ENTITY_STILL_ALIVE);
+     }
+ 
++    /** @deprecated Use IItemHandler capability instead. To preserve Container-specific interactions, use {@link #getContainerOrHandlerAt} and handle both cases. */
++    @Deprecated
```
Comment (sciwhiz12) [inline patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch:53]: Why was only this method deprecated? Why not the other `getContainerAt` overload, or `getSourceContainer`, or other nearby methods that work with `Container`?

File: src/main/java/net/neoforged/neoforge/items/ContainerOrHandler.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+package net.neoforged.neoforge.items;
+
+import net.minecraft.world.Container;
+import org.jetbrains.annotations.Nullable;
+
+public record ContainerOrHandler(
+        @Nullable Container container,
+        @Nullable IItemHandler itemHandler) {
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/items/ContainerOrHandler.java:13]: Should we enforce through the canonical constructor that either of the two components but not both must have a value? That's the implication of the name, at least.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch (hunk 2 above)
Comment (Technici4n) [reply patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch:53]: I only deprecated the public one, but I can deprecate all of them if you want.

File: src/main/java/net/neoforged/neoforge/items/ContainerOrHandler.java (hunk 3 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/items/ContainerOrHandler.java:13]: Yes, why not!

File: src/main/java/net/neoforged/neoforge/items/VanillaInventoryCodeHooks.java (hunk 1 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/items/VanillaInventoryCodeHooks.java]: OK, I replaced these two comments by an explanation of what each method does.

File: patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch (hunk 2 above)
Comment (sciwhiz12) [reply patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch:53]: Ah right, the other methods are `private`. Do we think it is likely that a mod might access-transform any of those methods to being `public` for their use? If not, we can just deprecate the `public` one and leave it at that.

Comment (Technici4n) [reply patches/net/minecraft/world/level/block/entity/HopperBlockEntity.java.patch:53]: I think it's unlikely that a modder would access-transform another of these methods without noticing that Minecraft doesn't use them anymore, and that the `public` `getContainerAt` overload is deprecated.

**Review by sciwhiz12 — APPROVED** [review]
